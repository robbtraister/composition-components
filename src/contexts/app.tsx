'use strict'

import React, { createContext, useContext } from 'react'

const appContext = createContext<{
  tree: object
  elements: object[]
  promises: any[]
  getContent?: (cp: ContentParams) => ContentResult
}>({
  tree: null,
  elements: [],
  promises: [],

  getContent: () => null
})

function useAppContext() {
  const { promises, ...consumableAppContext } = useContext(appContext)
  return consumableAppContext
}

function AppContext(props) {
  const { component: Component, render } = props
  const context = useAppContext()

  if (Component) {
    return <Component context={context} />
  } else if (render) {
    return render({ context })
  }
}

export default appContext
export { appContext, AppContext, useAppContext }
