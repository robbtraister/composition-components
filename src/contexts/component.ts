'use strict'

import { createContext, useContext } from 'react'

import { render, RenderableProps } from '../components/render'
import { NodeProps } from '../types'

const componentContext = createContext<NodeProps>({
  type: null,
  id: null,
  props: {}
})

export function useComponentContext() {
  const { customContent, ...consumableContext } = useContext(componentContext)
  return consumableContext
}

export function ComponentContext(props: RenderableProps<{}, NodeProps>) {
  const context = useComponentContext()
  return render({ ...props, ...context })
}

export default componentContext
