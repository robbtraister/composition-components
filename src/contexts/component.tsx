'use strict'

import React, { createContext, useContext } from 'react'

const componentContext = createContext<{
  type?: string
  id: any
  props?: object
  getContent?: (cp: ContentParams) => ContentResult
}>({
  type: null,
  id: null,
  props: {},

  getContent: () => null
})

export function useComponentContext() {
  return useContext(componentContext)
}

export function ComponentContext(props) {
  const { children, component: Component, render, ...otherProps } = props
  const context = useComponentContext()

  if (Component) {
    return <Component {...otherProps} context={context} />
  } else if (render) {
    return render({ ...otherProps, context })
  } else if (children) {
    return []
      .concat(children || [])
      .map((Child, index) => (
        <Child key={index} {...otherProps} context={context} />
      ))
  }
}

export default componentContext
