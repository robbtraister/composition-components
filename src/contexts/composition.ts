'use strict'

import { createContext, useContext } from 'react'

import { render } from '../components/render'

export const compositionContext = createContext<Composition.CompositionProps>(
  {}
)

export function useCompositionContext() {
  const {
    cache,
    quarantine,
    getComponent,
    getContent,
    getResource,
    ...consumableContext
  } = useContext(compositionContext)
  return consumableContext
}

export function CompositionContext(
  props: Composition.RenderableProps<{}, Composition.ContextStruct>
) {
  const context = useCompositionContext()
  return render({ ...props, context })
}

export default compositionContext
