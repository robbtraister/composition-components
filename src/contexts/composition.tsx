'use strict'

import React, { createContext, useContext } from 'react'

import { TreeNode } from '../components/tree'

const compositionContext = createContext<{
  cache?: object
  elements?: TreeNode[]
  pageContent?: object
  projectRoot?: string
  quarantine?: boolean
  tree?: TreeNode

  getComponent?: (node) => React.ComponentType
  getContent?: (cp: ContentParams) => ContentResult
}>({
  getComponent: () => React.Fragment,
  getContent: () => null
})

export function useCompositionContext() {
  const {
    cache,
    projectRoot,
    quarantine,
    ...consumableAppContext
  } = useContext(compositionContext)
  return consumableAppContext
}

export function CompositionContext(props) {
  const { component: Component, render } = props
  const context = useCompositionContext()

  if (Component) {
    return <Component context={context} />
  } else if (render) {
    return render({ context })
  }
}

export default compositionContext
