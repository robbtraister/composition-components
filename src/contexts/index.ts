'use strict'

export * from './app'
export * from './component'
