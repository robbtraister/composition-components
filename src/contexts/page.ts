'use strict'

import { createContext, useContext } from 'react'

import { render } from '../components/render'

const pageContext = createContext<Composition.PageProps>({})

export function usePageContext() {
  const {
    cache,
    quarantine,
    getComponent,
    getContent,
    getResource,
    ...consumableContext
  } = useContext(pageContext)
  return consumableContext
}

export function PageContext(
  props: Composition.RenderableProps<{}, Composition.PageProps>
) {
  const context = usePageContext()
  return render({ ...props, ...context })
}

export default pageContext
