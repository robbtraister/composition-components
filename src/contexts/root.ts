'use strict'

import { createContext, useContext } from 'react'

import { render, RenderableProps } from '../components/render'
import { RootProps } from '../components/root/types'

const rootContext = createContext<RootProps>({})

export function useRootContext() {
  const {
    cache,
    getComponent,
    getContent,
    getResource,
    ...consumableContext
  } = useContext(rootContext)
  return consumableContext
}

export function RootContext(props: RenderableProps<{}, RootProps>) {
  const context = useRootContext()
  return render({ ...props, ...context })
}

export default rootContext
