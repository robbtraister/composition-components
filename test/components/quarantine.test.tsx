'use strict'

/* global expect, test */

import React from 'react'
import { render } from '@testing-library/react'

import Composition from '../../src/components/composition'
import { useComponentContext } from '../../src/contexts/component'

const x = 3
function Fail() {
  if (x === 3) {
    throw new Error('component render failed')
  }
  return <React.Fragment />
}

function Success({ children }: { children: React.ElementType }) {
  const { id, type } = useComponentContext()
  return (
    <div data-type={type} data-id={id}>
      {children}
    </div>
  )
}

function getComponent(type: string): React.ComponentType {
  return type === 'fail' ? Fail : Success
}

const tree = {
  type: 'div',
  id: 'abc',
  children: [
    {
      type: 'fail',
      id: 'def',
      children: [
        {
          type: 'div',
          id: 'ghi'
        }
      ]
    },
    {
      type: 'div',
      id: 'xyz'
    }
  ]
}

test('Quarantine Component', () => {
  const { asFragment } = render(
    <Composition getComponent={getComponent} tree={tree} quarantine />
  )
  expect(asFragment()).toMatchSnapshot()
})
