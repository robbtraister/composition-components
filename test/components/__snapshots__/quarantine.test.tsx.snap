// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quarantine Component 1`] = `
<DocumentFragment>
  <div
    data-id="abc"
    data-type="div"
  >
    <div
      data-error-component="fail"
      data-error-id="def"
      data-error-message="component render failed"
      style="display: none;"
    />
    <div
      data-id="xyz"
      data-type="div"
    />
  </div>
</DocumentFragment>
`;
